### ISSUE TITLE
Multiple host viewports

### ISSUE THREAD

**Issue by epajarre** [ocornut/imgui#3012]
**Version/Branch of Dear ImGui:**

Version: V1.75 (current GIT)
Branch: started with master, now looking at docking

**Back-end/Renderer/Compiler/OS**

Back-ends: imgui_impl_opengl2.cpp + imgui_impl_glfw.cpp
Compiler: Visualstudio (+gcc/clang)
Operating System:  Windows 10 (Windows as primary, also Linux)

**My Issue/Question:**

I am experimenting with ImGui with a rather old code base, I would like to migrate imy program from the Gwen user interface. The program uses multiple displays, and typically has an window (currently Glfw window) for each display. I need to show a GUI on two of the displays.

I do not need to move ImGui windows from their original host windows, and the GUI  elements on different displays are actually independent. The host windows can be either in normal or fullscreen windows.  At the moment I am not using threads (well the GUI operation happens in the main thread)

I originally tried to solve my problem using separate ImGui contexts for the two GUIs, but I think there was some issues with global/static shared state even when using different contexts. The Docking branch has support for multiple windows when additional platform windows are created by ImGui, but should it work with multiple host windows?

I guess I could make this work by using namespaces and integrating ImGui twice to my software. I originally thought that this would be a "dirty" solution, but I am not so sure any more.

So I would appreciate any pointers/examples regarding how I should go forward. Master vs Docking branch? Single context or swapping Context? Or maybe just take the namespace route for now? Issues   #586 and #269 seem to be related, but I am not quite sure how much ImGui has changed since they were created.

And my apologies if I just did not find an already existing example. (and also for not providing example code from my attempts). 

 Eero

**Comment by ocornut** [comment]
Hello Eero,

It looks like you looked in the right spots already and have enough understanding of what's available.

If you are not using multiple threads for the UI I don't understand what's the problem with using multiple contexts? You can just use `SetCurrentContext()` to swap them and do your stuff. If you absolutely need to emit UI from multiple threads you can either threat the "current context" as a resource to mutex/lock into, either `#define GImGui` into a thread-local variable so each others can point to a different value (with a bit of extra overhead accessing that variable). 

Multi-viewports available in the `docking` branch (check Wiki>Glossary if you are not sure about the difference) currently don't support multiple host windows. They will eventually but that shouldn't really stop you: if the imgui windows become their own host windows rather than being hosted in existing windows there isn't much difference from a user point of view.?

**Comment by epajarre** [comment]
Ok, thanks!

I will try again with the SetCurrentContext() approach and the Master branch. I will add notes here, if I find anything interesting. I have some schedule issues, so the namespace solution will be my plan B if there are significant problems.

At the moment I prefer keeping my ImGui windows embedded inside the host windows as the typical end target is doing several full screen views, (or sometimes something embedded in a VR view). Also I seem to have significant performance issues with the window/context swapping anyways. One of the reasons going to ImGui is possible migration from OpenGL to Vulkan.

**Comment by epajarre** [comment]
Been a while, but I think I should report my experiences with using Imgui (main branch), with two or more toplevel Opengl (old style) + glfw windows. 

1.  There has been no reason to modify the main ImGui sources, the work as they are.

2. For my secondary glfw windows I pass the primary window as a parameter for glfwCreateWindow so that I can share textures (fonts).

3. I pass false for install_callbacks parameter to ImGui_ImplGlfw_InitForOpenGL function, and setup the callbacks manually, also I only call the Init function for the first window. For the secodary window I create the ImGui::Context with CreateContext while passing the io.Fonts from the first ImGui context.

4.  I have done some changes to examples/imgui_impl_glfw

I added GLFWwindow * parameter for ImGui_ImlGlfw_NewFrame, so I can use the correct window for the Window/frame buffer calls and pass this parameter to the ImGui_ImplGLfw_UpdateMouse...() functions.  Maybe this could have been done by integrating GLFWwindow* to the ImGui context, but I thought my change was less invasive.

Also  in imgui_impl_glfw.cpp I changed the g_MouseJustPressed array from bool to GLFWindow * type. This way I can keep note, on which window the press happened and only use and reset the flag in ImGui_ImplGLfw_UpdateMouseAndButtons if we are processing that window.


I propably have missed some things, but so far it seems to work.

 Eero
